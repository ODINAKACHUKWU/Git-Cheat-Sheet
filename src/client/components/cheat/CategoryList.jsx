import React from 'react';
import PropTypes from 'prop-types';
import CopyToClipboard from 'react-copy-to-clipboard';
import Masonry from 'react-masonry-component';

const CategoryList = ({
  allGitCheats,
  toggleDisplay,
  isDisplayed,
  cheatId
}) => (
  <Masonry className="card-columns">
    {allGitCheats.map(category => (
      <div className="col-md-4" key={category.name}>
        <div className="card">
          <div className="card-header">{category.name}</div>
          {category.cheats.map(cheat => (
            <div key={cheat._id} className="pt-0 pr-2 pl-2 pb-2">
              <small className="card-title">{cheat.description}</small>
              <small
                style={{
                  display: `${
                    isDisplayed && cheatId === cheat._id ? 'inline' : 'none'
                  }`,
                  float: 'right',
                  fontStyle: 'italic',
                  fontSize: '10px'
                }}
              >
                Copied!
              </small>
              <small>
                <CopyToClipboard
                  text={cheat.command}
                  onCopy={() => toggleDisplay(cheat._id)}
                >
                  <code className="card-text d-block">
                    <span className="mr-2">$</span>
                    {cheat.command}
                  </code>
                </CopyToClipboard>
              </small>
            </div>
          ))}
        </div>
      </div>
    ))}
  </Masonry>
);
CategoryList.propTypes = {
  allGitCheats: PropTypes.array.isRequired,
  toggleDisplay: PropTypes.func.isRequired,
  isDisplayed: PropTypes.bool.isRequired,
  cheatId: PropTypes.string.isRequired
};

export default CategoryList;
