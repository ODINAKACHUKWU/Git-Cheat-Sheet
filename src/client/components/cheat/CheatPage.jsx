import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { MDBRow, MDBCol } from 'mdbreact';
import Spinner from 'react-md-spinner';

import NavBar from '../common/NavBar';
import SearchBar from '../common/SearchBar';
import CategoryList from './CategoryList';
import getCheats, { searchCheats } from '../../actions/creators/cheatsAction';
import { logout } from '../../actions/creators/userAuthActions';

export class CheatPage extends Component {
  state = {
    isDisplayed: false,
    cheatId: '0',
    searchText: ''
  };

  componentDidMount() {
    const { getCheats } = this.props;
    getCheats();
  }

  handleToggleDisplay = cheatId => {
    this.setState({ isDisplayed: true, cheatId }, () => {
      setTimeout(() => this.setState({ isDisplayed: false }), 1500);
    });
  };

  handleLogout = () => {
    const { logout, history } = this.props;
    logout();
    history.push('/signin');
  };

  handleOnChange = event => {
    event.persist();
    this.setState(
      prevState => ({
        ...prevState,
        searchText: event.target.value
      }),
      () => this.handleCheatSearch()
    );
  };

  handleCheatSearch = () => {
    const { searchText } = this.state;
    const { searchCheats } = this.props;
    searchCheats(searchText);
  };

  render() {
    const { isFetching, allGitCheats, searchData, username } = this.props;
    const { isDisplayed, cheatId } = this.state;

    const cheatsData = searchData || allGitCheats;

    return (
      <MDBRow className="d-flex justify-content-center">
        <MDBCol md="10">
          <NavBar username={username} logout={this.handleLogout} />
          {isFetching ? (
            <div
              style={{
                display: 'flex',
                justifyContent: 'center',
                marginTop: '20%'
              }}
            >
              <Spinner />
            </div>
          ) : (
            <Fragment>
              <SearchBar onChangeHandler={this.handleOnChange} />
              <CategoryList
                allGitCheats={cheatsData}
                isDisplayed={isDisplayed}
                toggleDisplay={this.handleToggleDisplay}
                cheatId={cheatId}
              />
            </Fragment>
          )}
        </MDBCol>
      </MDBRow>
    );
  }
}

const mapStateToProps = ({
  gitCheats: { isFetching, allGitCheats, searchData },
  userAuth: {
    user: { username }
  }
}) => ({
  isFetching,
  allGitCheats,
  searchData,
  username
});

CheatPage.propTypes = {
  isFetching: PropTypes.bool.isRequired,
  allGitCheats: PropTypes.array.isRequired,
  getCheats: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
  username: PropTypes.string,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  searchCheats: PropTypes.func.isRequired
};

CheatPage.defaultProps = {
  username: 'username'
};

export default connect(
  mapStateToProps,
  { getCheats, logout, searchCheats }
)(CheatPage);
