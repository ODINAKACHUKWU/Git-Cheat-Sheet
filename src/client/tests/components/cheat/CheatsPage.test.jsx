import { CheatPage } from '../../../components/cheat/CheatPage';

describe('<SignInPage /> component', () => {
  const setUp = () => {
    const props = {
      isFetching: false,
      allGitCheats: [
        {
          name: 'Install Git',
          cheats: [{}]
        }
      ],
      getCheats: jest.fn(() => Promise.resolve()),
      logout: jest.fn(),
      username: 'some_user',
      history: {
        push: jest.fn()
      },
      searchCheats: jest.fn()
    };
    const state = {
      isDisplayed: false,
      cheatId: '0',
      searchText: ''
    };
    return {
      shallowWrapper: shallow(<CheatPage {...props} />),
      state,
      props
    };
  };

  const event = {
    target: {
      searchText: 'push'
    },
    preventDefault: jest.fn(),
    persist: jest.fn()
  };

  it('render without crashing', () => {
    const { shallowWrapper } = setUp();
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should display <Loader /> component when `isFetching` is set to true', () => {
    const { shallowWrapper, props } = setUp();
    shallowWrapper.setProps({
      ...props,
      isFetching: true
    });

    expect(shallowWrapper.find('MDSpinner')).toHaveLength(1);
    expect(shallowWrapper.find('SearchBar')).toHaveLength(0);
    expect(shallowWrapper.find('CategoryList')).toHaveLength(0);
  });

  it('invokes componentDidMount method', () => {
    const { shallowWrapper, props } = setUp();
    const componentDidMountSpy = jest.spyOn(
      shallowWrapper.instance(),
      'componentDidMount'
    );
    shallowWrapper.instance().componentDidMount();

    expect(componentDidMountSpy).toHaveBeenCalled();
    expect(props.getCheats).toHaveBeenCalled();
  });

  it('invokes handleOnChange method', () => {
    const { shallowWrapper } = setUp();
    const handleOnChangeSpy = jest.spyOn(
      shallowWrapper.instance(),
      'handleOnChange'
    );
    shallowWrapper.instance().handleOnChange(event);
    expect(handleOnChangeSpy).toHaveBeenCalledWith(event);
  });

  it('invokes handleToggleDisplay method', () => {
    const { shallowWrapper } = setUp();
    const handleToggleDisplaySpy = jest.spyOn(
      shallowWrapper.instance(),
      'handleToggleDisplay'
    );
    shallowWrapper.instance().handleToggleDisplay('cheatId');

    expect(handleToggleDisplaySpy).toHaveBeenCalledWith('cheatId');
  });

  it('invokes handleCheatSearch method', () => {
    const { props } = setUp();
    const mountWrapper = mount(<CheatPage {...props} />);

    mountWrapper
      .find('SearchBar')
      .find('input')
      .simulate('change');
    expect(props.searchCheats).toHaveBeenCalled();
  });

  it('should logout user', () => {
    const { props } = setUp();
    const mountWrapper = mount(<CheatPage {...props} />);

    mountWrapper
      .find('NavbarPage')
      .find('button')
      .simulate('click');

    expect(props.logout).toHaveBeenCalled();
  });
});
