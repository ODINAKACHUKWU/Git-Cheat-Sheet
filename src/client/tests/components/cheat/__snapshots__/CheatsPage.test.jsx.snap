// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignInPage /> component render without crashing 1`] = `
<t
  className="d-flex justify-content-center"
  tag="div"
>
  <t
    lg={null}
    md="10"
    sm={null}
    tag="div"
    xl={null}
    xs={null}
  >
    <NavbarPage
      logout={[Function]}
      username="some_user"
    />
    <SearchBar
      onChangeHandler={[Function]}
    />
    <CategoryList
      allGitCheats={
        Array [
          Object {
            "cheats": Array [
              Object {},
            ],
            "name": "Install Git",
          },
        ]
      }
      cheatId="0"
      isDisplayed={false}
      toggleDisplay={[Function]}
    />
  </t>
</t>
`;
